node/src/components/small_network/tasks.rs

```diff
@@ -0,0 +1,478 @@
+    let mut tls_stream = tls::create_tls_acceptor(&cert.as_x509().as_ref(), &secret_key.as_ref())
+        .and_then(|ssl_acceptor| Ssl::new(ssl_acceptor.context()))
+        .and_then(|ssl| SslStream::new(ssl, stream))
+        .map_err(ConnectionError::TlsInitialization)?;
+
+    SslStream::accept(Pin::new(&mut tls_stream))
+        .await
+        .map_err(ConnectionError::TlsHandshake)?;
+
+    // We can now verify the certificate.
+    let peer_cert = tls_stream
+        .ssl()
+        .peer_certificate()
+        .ok_or(ConnectionError::NoPeerCertificate)?;
+
+    Ok((
+        NodeId::from(
+            tls::validate_cert(peer_cert)
+                .map_err(ConnectionError::PeerCertificateInvalid)?
+                .public_key_fingerprint(),
+        ),
+        tls_stream,
+    ))
+}
+
+/// Performs an IO-operation that can time out.
+async fn io_timeout<F, T, E>(duration: Duration, future: F) -> Result<T, IoError<E>>
+where
+    F: Future<Output = Result<T, E>>,
+    E: StdError + 'static,
+{
+    tokio::time::timeout(duration, future)
+        .await
+        .map_err(|_elapsed| IoError::Timeout)?
+        .map_err(IoError::Error)
+}
+
+/// Performs an IO-operation that can time out or result in a closed connection.
+async fn io_opt_timeout<F, T, E>(duration: Duration, future: F) -> Result<T, IoError<E>>
+where
+    F: Future<Output = Option<Result<T, E>>>,
+    E: StdError + 'static,
+{
+    let item = tokio::time::timeout(duration, future)
+        .await
+        .map_err(|_elapsed| IoError::Timeout)?;
+
+    match item {
+        Some(Ok(value)) => Ok(value),
+        Some(Err(err)) => Err(IoError::Error(err)),
+        None => Err(IoError::UnexpectedEof),
+    }
+}
+
+async fn negotiate_handshake<P, REv>(
+    context: &NetworkContext<REv>,
+    transport: &mut FramedTransport<P>,
+) -> Result<SocketAddr, ConnectionError>
+where
+    P: Payload,
+{
+    // Send down a handshake and expect one in response.
+    let handshake = context.chain_info.create_handshake(context.public_addr);
+
+    io_timeout(HANDSHAKE_TIMEOUT, transport.send(handshake))
+        .await
+        .map_err(ConnectionError::HandshakeSend)?;
+
+    let remote_handshake = io_opt_timeout(HANDSHAKE_TIMEOUT, transport.next())
+        .await
+        .map_err(ConnectionError::HandshakeRecv)?;
+
+    if let Message::Handshake {
+        network_name,
+        public_addr,
+        protocol_version,
+    } = remote_handshake
+    {
+        debug!(%protocol_version, "handshake received");
+
+        // The handshake was valid, we can check the network name.
+        if network_name != context.chain_info.network_name {
+            Err(ConnectionError::WrongNetwork(network_name))
+        } else {
+            Ok(public_addr)
+        }
+    } else {
+        // Received a non-handshake, this is an error.
+        Err(ConnectionError::DidNotSendHandshake)
+    }
+}
+
+/// Core accept loop for the networking server.
+pub(super) async fn server<P, REv>(
```

Comment:
A verb would be more appropriate - `serve` or `run_server` maybe?

Reply:
This was intentional, the noun-named tasks are long-running until terminated, while the verb functions/tasks are self-terminating. `server`, `message_sender` and `message_receiver` are in the former category. Maybe it's worth keeping this little scheme?

Reply:
It was a nitpick, so since there's a rationale for it, I'm fine to leave it as is.